docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,540 @@
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+#### What is being proposed
+
+Design of a subsystem relevant in the multi-tenant setting ("serverless") which
+rate limits tenant KV operations in conformance to a budget target.
+
+#### Why (short reason)
+
+Serverless users will pay per usage and need to be able to set an upper bound on
+how much they are willing to pay in a billing cycle. Free tier users must be
+rate limited in order to curtail resource usage.
+
+#### How (short plan)
+
+To encompass the various resources used, we have created an abstract unit called
+a "request unit". There will be a direct conversion from RUs to a dollar amount
+that is part of a tenant's bill. Request units are determined based on CPU usage
+in the SQL pods and on the KV requests from the tenant. For more details, see
+the "Managing Serverless Resource Usage" RFC
+([internal-only](https://github.com/cockroachlabs/managed-service/blob/master/docs/RFCS/20210603_serverless_usage.md)).
+
+Each tenant will have a number of RUs available to use as needed, plus a
+"baseline" rate of replenishment. These map directly to the concept of a
+[token bucket](https://en.wikipedia.org/wiki/Token_bucket).
+The values are determined by a higher-level billing algorithm that is beyond the
+scope of this RFC and need to be configurable on-the-fly.
+
+The enforcement of RU consumption will happen inside the SQL pods, at the point
+where KV requests are sent. The SQL pods communicate with centralized state (for
+that tenant) which lives in a system table inside the system tenant. The
+centralized state also tracks and reports the consumption of RUs.
+
+#### Impact
+
+The new subsystem will only be enabled on our multi-tenant clusters. The
+subsystem involves overhead related to communicating with and updating the
+centralized state for each tenant.
+
+Note that admission control for preventing overload in the host cluster or for
+reducing the variability in user-visible performance (depending on other load on
+the host cluster) are not in the scope of this RFC.
+
+# Motivation
+
+The higher-level motivation is covered by the "Managing Serverless Resource
+Usage" RFC ([internal-only](https://github.com/cockroachlabs/managed-service/blob/master/docs/RFCS/20210603_serverless_usage.md)).
+This RFC covers the implementation details of a subsystem necessary to enforce
+resource usage.
+
+# Technical design
+
+## Overview
+
+We implement a "global" (distributed) token bucket for each tenant; the metadata
+for all these buckets lives in a system table on the system tenant. The global
+bucket is configured with:
+ - an initial amount of "burst" tokens (RUs)
+ - a refill rate (RUs/second)
+ - a limit on how many unused tokens (RUs) we can accumulate as burst. Refill is
+   essentially paused when the bucket has more tokens than the limit.
+
+Each SQL pod implements a local token bucket and uses it for admission control
```

Comment:
I understand that this RFC does not cover overload admission control, but I am interested in the intersection between the 2 subsystems.
* What happens if a SQL tenant uses tokens to only be rejected by the overload admission control?